[BUG] agent: Channel selection request should yield a single channel selection response

VladyslavTupikin opened the pull request:
Update channel selection according to the UML diagram

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4319,6 +4328,154 @@ bool backhaul_manager::add_link_metrics(const sMacAddr &reporter_al_mac,
+            ret = true;
+            break;
+        }
+    }
```
All this is not needed, because we expect a response from *all* slaves. So what we need to do is simply to populate m_channel_selection_request with all slaves. And it has to be forwarded to the slaves, but for that we can simply return false.

VladyslavTupikin replied:
Fixed

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4319,6 +4328,154 @@ bool backhaul_manager::add_link_metrics(const sMacAddr &reporter_al_mac,
+    //           << int(channel_selection_response->response_code());
+
+    m_channel_selection_response.push_back({channel_selection_response->radio_uid(),
+                                            uint8_t(channel_selection_response->response_code())});
```
(nitpick) Thanks to @mariomaz, I discovered that we can use emplace_back instead of push_back. It avoids an additional copy.

VladyslavTupikin replied:
emphase_back does not work for the marked case.
It will be work if i add 3 useless strings for creating and filling structure elements, then structure object i will be able to put inside the vector via amphase_back

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -552,6 +556,14 @@ class backhaul_manager : public btl::transport_socket_thread {
+    };
+
+    std::vector<sMacAddr> m_channel_selection_request;
+    std::vector<sChannelSelectionResponse> m_channel_selection_response;
```
We should also store the MID, and check it when collecting responses.
Therefore, I think it would be nicer like this:
```suggestion
   struct sExpectedChannelSelection {
       uint16_t mid;
       std::vector<sMacAddr> requested;
       std::vector<sChannelSelectionResponse> responses;
   };
   sExpectedChannelSelection m_expected_channel_selection;
```

VladyslavTupikin replied:
Fixed.

arnout reviewed (dismissed):
Although the changes aren't that big, I'd like to re-review.

Also, still two things are missing:
- Final commit should have the magic message to run the 4.5.2 test.
- test_flows should be updated: there are a few TODOs in the channel selection that are now fixed.

In fact, the test flows stuff is how you should test if the issue is really fixed. So it's even better to first do the test_flows changes and then modify the code so the test passes.

Commit:
agent: backhaul_manager: Update channel selection

Update channel selection according to the UML diagram.

Add capture channel selection query message
by backhaul manager, save message id and radio mac,
forward message tp the sun_slaves. Collect
channel selection responses and send 1 complete
channel selection response to the controller.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4319,6 +4328,85 @@ bool backhaul_manager::add_link_metrics(const sMacAddr &reporter_al_mac,
+
+    LOG(DEBUG) << "Forwarding CHANNEL_SELECTION_REQUEST to son_slave, mid=" << std::hex << int(mid);
+    m_expected_channel_selection.mid = mid;
+    for (auto &socket : slaves_sockets) {
```
If you are overwriting MID value, then I think you should first clear the list

mariomaz replied:
And also clear the list of responses

VladyslavTupikin replied:
In the `handle_slave_channel_selection_response` method `requests` vector erases, then if `requests` vector empty, response message will be prepared for the controller. Before sending the response to the controller,  `responses` vector erases too.

mariomaz replied:
Sorry, I didn't explain myself properly: I meant you have to consider the case when agent receives a CHANNEL_SELECTION_REQUEST **before** it has completely processed a previous one.

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4319,6 +4328,85 @@ bool backhaul_manager::add_link_metrics(const sMacAddr &reporter_al_mac,
+    for (auto &socket : slaves_sockets) {
+        m_expected_channel_selection.requests.emplace_back(socket->radio_mac);
+    }
+    // Forward to all slaves
```
This is quite dangerous since we are handling this message by doing nothing and having faith that the default error handling will do the job for us. IIRC such default handling is going to change in a near future and this code will stop working.
Anyway, I think the comment should better explain what's going on here so maintainers can fix it if needed.

VladyslavTupikin replied:
Added advanced comment message in that place.

arnout replied:
> This is quite dangerous since we are handling this message by doing nothing and having faith that the default error handling will do the job for us. IIRC such default handling is going to change in a near future and this code will stop working.

Yes it will change, but in the best way possible: there will be no slaves any more, so no forwarding needs to be done at all.

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4319,6 +4328,85 @@ bool backhaul_manager::add_link_metrics(const sMacAddr &reporter_al_mac,
+        return false;
+    }
+
+    for (auto &response : m_expected_channel_selection.responses) {
```
```suggestion
    for (const auto &response : m_expected_channel_selection.responses) {
```

VladyslavTupikin replied:
fixed

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -552,6 +556,18 @@ class backhaul_manager : public btl::transport_socket_thread {
 
+    struct sChannelSelectionResponse {
+        sMacAddr radio_mac;
+        uint8_t response_code;
```
```suggestion
         wfa_map::tlvChannelSelectionResponse::eResponseCode response_code;
```

With this change you save two type-casts in .cpp file

VladyslavTupikin replied:
fixed

mariomaz reviewed (changes requested):
Not approving because there are some changes to do

Commit:
tests: test_flows.py: Fix channel selection test

Need to fix 'TODO: should be a single response'
for channel selection response.

Add new method, which cheks channel selection
message and tlv payload.

MAP-4.5.2_ETH_FH5GL:netgear-rax40

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4319,6 +4327,89 @@ bool backhaul_manager::add_link_metrics(const sMacAddr &reporter_al_mac,
+    const auto mid = cmdu_rx.getMessageId();
+
+    LOG(DEBUG) << "Forwarding CHANNEL_SELECTION_REQUEST to son_slave, mid=" << std::hex << int(mid);
+    m_expected_channel_selection.mid = mid;
```
As @mariomaz said, there should be something like this:

```suggestion
    // Clear previous request, if any
    m_expected_channel_selection.requests.clear();
    m_expected_channel_selection.responses.clear();
    
    m_expected_channel_selection.mid = mid;
```

VladyslavTupikin replied:
added proposed fix into the code
